### PR TITLE
✨ Source MongoDB Internal POC: CDC Metadata Injector

### PR DESCRIPTION
**Pull request by jdpgrailsdev**
## What
* Inject Airbyte-required metadata for CDC mode in MongoDB source

## How
* Implement the `CdcMetadataInjector` interface for MongoDB
* Refactor the `DebeziumEventUtils` and `CdcMetadataInjector` implementation to push the retrieval of the stream name down into the metadata injector to support the difference between the relational database Debezium connectors and MongoDB.

## Recommended reading order
1. `CdcMetadataInjector.java`
2. `MongoDbCdcConnectorMetadataInjector.java`
3. `DebeziumEventUtils.java`

**N.B.**: This PR is based off of https://github.com/airbytehq/airbyte/pull/29710, as it depends on some of the utility methods for dealing with resume tokens and timestamps.

### CODE REVIEW

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * MongoDB specific implementation of the {@link CdcMetadataInjector} that stores the MongoDB resume
+ * token timestamp in the {@link DebeziumEventUtils#CDC_LSN} metadata field.
+ */
+public class MongoDbCdcConnectorMetadataInjector implements CdcMetadataInjector<BsonTimestamp> {
+
+  @Override
+  public void addMetaData(final ObjectNode event, final JsonNode source) {
+    final BsonTimestamp timestamp = MongoDbResumeTokenHelper.extractTimestampFromSource(source);
+    event.put(CDC_LSN, timestamp.getValue());
```
Comment (subodh1810) [inline airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:28]: Mongo has a concept of LSN?

**Review by subodh1810 — COMMENTED** [review]
Are we sure about the metadata columns that we want to add to records, Mongo has a concept of LSN?

**Review by jpefaur — APPROVED** [review]

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  public MongoDbCdcConnectorMetadataInjector(long emittedAtConverted) {
+    this.emittedAtConverted = emittedAtConverted;
+  }
+
+  @Override
+  public void addMetaData(final ObjectNode event, final JsonNode source) {
+    final BsonTimestamp timestamp = MongoDbResumeTokenHelper.extractTimestampFromSource(source);
+    event.put(CDC_UPDATED_AT, timestamp.getValue());
+    event.put(CDC_DEFAULT_CURSOR, getCdcDefaultCursor());
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:41]: This is used by the V2 destination to track the order of records.  The implementation here is copied from the MySql injector to guarantee a globally unique, increasing value.  We could have potentially used the _id value, but there is no guarantee that if you sort the string value that the order matches the natural document order, so you would probably need to decode it, if possible, into a timestamp.  Therefore, we decided to go with this approach that is already in use by the other CDC sources.

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  @Override
+  public void addMetaDataToRowsFetchedOutsideDebezium(final ObjectNode record, final String transactionTimestamp, final BsonTimestamp metadataToAdd) {
+    record.put(CDC_UPDATED_AT, transactionTimestamp);
+    record.put(CDC_DELETED_AT, (String) null);
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:47]: I believe that this is to have parity with the incremental sync side of things, which may actual delete records.  This code is in the method called as part of the initial snapshot, so none of those records would ever be deleted.

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    return source.get(MongoDbDebeziumConstants.ChangeEvent.SOURCE_DB).asText();
+  }
+
+  @Override
+  public String name(JsonNode source) {
+    return source.get(MongoDbDebeziumConstants.ChangeEvent.SOURCE_COLLECTION).asText();
+  }
+
+  private Long getCdcDefaultCursor() {
+    return this.emittedAtConverted + this.recordCounter.getAndIncrement();
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:62]: It supposedly is generating an unique, always increasing value for use by the V2 to source to order the records.

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  private final AtomicLong recordCounter = new AtomicLong(1);
+
+  public MongoDbCdcConnectorMetadataInjector(long emittedAtConverted) {
+    this.emittedAtConverted = emittedAtConverted;
+  }
+
+  @Override
+  public void addMetaData(final ObjectNode event, final JsonNode source) {
+    final BsonTimestamp timestamp = MongoDbResumeTokenHelper.extractTimestampFromSource(source);
+    event.put(CDC_UPDATED_AT, timestamp.getValue());
```
Comment (subodh1810) [inline airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:50]: We are already adding `base.put(CDC_UPDATED_AT, transactionTimestamp);` This is redundant right?

**Review by subodh1810 — CHANGES_REQUESTED** [review]
I pushed a commit to make the class singleton like Mysql one. Also a comment around adding the `CDC_UPDATED_AT` again

**Comment by subodh1810** [comment]
Also lets run tests for all the CDC connectors

File: airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java (hunk 5 above)
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/source-mongodb-internal-poc/src/main/java/io/airbyte/integrations/source/mongodb/internal/cdc/MongoDbCdcConnectorMetadataInjector.java:50]: @subodh1810 I didn't realize that the `event` passed to `addMetaData` contains the metadata from the other method.  If that is the case, then it is redundant, though I think we then need to talk about which value is the correct one.

**Review by subodh1810 — APPROVED** [review]
Looks good! Lets make sure the tests for Postgres, mysql and mssql pass before merging as well.

**Comment by octavia-approvington** [comment]
It's time <br/> fine lets go

**Review by octavia-approvington — APPROVED** [review]

**Pull request merged**